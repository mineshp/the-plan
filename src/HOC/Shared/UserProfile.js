import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { retrieveProfiles } from '../../actions/controlCentre';
import { addNotification } from '../../actions/notification';
import { setProfilesToDisplay, setCurrentUser } from '../../actions/authentication';
import UserProfileComponent from '../../components/Shared/UserProfile';

import LoadingComponent from '../../components/Shared/Loading';

/* istanbul ignore next: not testing buildProjectDropdownOptions */
const buildProfileDropdownOptions = (profiles) => {
    if (!profiles) { return []; }
    return profiles.map((profile) => ({
        // eslint-disable-next-line no-underscore-dangle
        key: profile._id,
        value: profile.name,
        text: profile.name
    }));
};

class UserProfile extends Component {
    constructor(props, context) {
        super(props, context);

        this.handleDropDownSelection =
            this.handleDropDownSelection.bind(this);

        this.handleSubmit = this.handleSubmit.bind(this);
        this.state = { profilesAssigned: this.props.user.profilesToDisplay };
    }

    async componentDidMount() {
        await this.props.actions.retrieveProfiles();
    }

    handleDropDownSelection(event, data) {
        event.preventDefault();
        this.setState({ profilesAssigned: [...data.value] });
    }

    handleSubmit(event) {
        event.preventDefault();
        this.props.actions.setProfilesToDisplay(this.state.profilesAssigned, this.props.user)
            .then(() => {
                this.props.actions.addNotification(this.props.notification);
                this.context.router.history.push('/');
            });
    }

    /* istanbul ignore next: not testing render */
    render() {
        const { admin, user } = this.props;
        const userProfiles = admin.controlCentre.profiles.filter((elem) =>
            user.profile.find((profile) => profile === elem.name));

        return (
            !admin.controlCentre || !admin.controlCentre.profiles || !userProfiles
                ? <LoadingComponent />
                : <UserProfileComponent
                    profileOptions={buildProfileDropdownOptions(userProfiles)}
                    profilesAssigned={this.state.profilesAssigned}
                    handleDropDownSelection={this.handleDropDownSelection}
                    handleSubmit={this.handleSubmit}
                />
        );
    }
}

/* istanbul ignore next: not testing mapStateToProps */
const mapStateToProps = (state) => ({
    admin: state.controlCentre,
    user: state.authentication.user,
    notification: state.authentication.notification,
});

/* istanbul ignore next: not testing mapDispatchToProps */
const mapDispatchToProps = (dispatch) => (
    {
        actions: bindActionCreators({
            retrieveProfiles, setProfilesToDisplay, addNotification, setCurrentUser
        }, dispatch)
    }
);

const UserProfileConnectedComponent = connect(
    mapStateToProps,
    mapDispatchToProps
)(UserProfile);

UserProfile.propTypes = {
    user: PropTypes.shape({
        id: PropTypes.string.isRequired,
        username: PropTypes.string,
        isAdmin: PropTypes.bool,
        profile: PropTypes.arrayOf(PropTypes.string),
        profilesToDisplay: PropTypes.arrayOf(PropTypes.string)
    }),
    actions: PropTypes.shape({
        retrieveProfiles: PropTypes.func.isRequired,
        setProfilesToDisplay: PropTypes.func.isRequired,
        setCurrentUser: PropTypes.func.isRequired,
        addNotification: PropTypes.func.isRequired
    }),
    admin: PropTypes.shape({
        controlCentre: PropTypes.shape({
            profiles: PropTypes.arrayOf(PropTypes.shape({}))
        })
    }),
    notification: PropTypes.shape({
        message: PropTypes.string,
        level: PropTypes.string,
        title: PropTypes.string
    })
};

UserProfile.defaultProps = {
    user: null,
    actions: null,
    admin: null,
    notification: null
};

UserProfile.contextTypes = {
    router: PropTypes.object
};

export {
    UserProfile,
    UserProfileConnectedComponent
};
