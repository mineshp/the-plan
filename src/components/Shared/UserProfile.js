import React from 'react';
import PropTypes from 'prop-types';
import { Button, Container, Form, Header, Select } from 'semantic-ui-react';

const UserProfile = ({
    profilesAssigned,
    profileOptions,
    handleDropDownSelection,
    handleSubmit
}) => (
    <div className="List main content-body">
        <Container>
            <Header as="h1">Select Profile(s) to access</Header>
            <p>
                This sets the profiles available for you to access, the dropdown provides all possible profiles you have access to. All projects are assigned to a profile, so if you do not select the correct profiles you may not see what you are expecting.
            </p>
            <Form onSubmit={handleSubmit}>
                <Select
                    placeholder="Profiles"
                    fluid
                    multiple
                    search
                    selection
                    options={profileOptions}
                    onChange={handleDropDownSelection}
                    defaultValue={profilesAssigned}
                />
                <hr />
                <Button color="green" type="submit">Save</Button>
            </Form>

        </Container>
    </div>
);

UserProfile.propTypes = {
    handleDropDownSelection: PropTypes.func.isRequired,
    handleSubmit: PropTypes.func.isRequired,
    profileOptions: PropTypes.arrayOf(PropTypes.shape({})),
    profilesAssigned: PropTypes.arrayOf(PropTypes.string)
};

UserProfile.defaultProps = {
    profileOptions: [],
    profilesAssigned: []
};

export default UserProfile;

