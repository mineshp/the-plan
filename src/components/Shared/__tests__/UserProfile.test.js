import React from 'react';
import renderer from 'react-test-renderer';
import UserProfile from '../UserProfile';
import { mockUser, mockProfileOptions } from '../../../helpers/test/testData/authenticationData';

it('renders correctly', () => {
    const tree = renderer.create(<UserProfile
        profileOptions={mockProfileOptions()}
        profilesAssigned={mockUser().profile}
        handleDropDownSelection={jest.fn()}
        handleSubmit={jest.fn()}
    />).toJSON();
    expect(tree).toMatchSnapshot();
});
