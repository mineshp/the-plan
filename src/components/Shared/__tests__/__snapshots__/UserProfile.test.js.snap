// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="List main content-body"
>
  <div
    className="ui container"
  >
    <h1
      className="ui header"
    >
      Select Profile(s) to access
    </h1>
    <p>
      This sets the profiles available for you to access, the dropdown provides all possible profiles you have access to. All projects are assigned to a profile, so if you do not select the correct profiles you may not see what you are expecting.
    </p>
    <form
      action={undefined}
      className="ui form"
      onSubmit={[Function]}
    >
      <div
        aria-busy={undefined}
        aria-disabled={undefined}
        aria-expanded={false}
        className="ui fluid multiple search selection dropdown"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        role="combobox"
        tabIndex={undefined}
      >
        <a
          className="ui label"
          onClick={[Function]}
          value="TESTA"
        >
          TESTA
          <i
            aria-hidden="true"
            className="delete icon"
            onClick={[Function]}
          />
        </a>
        <a
          className="ui label"
          onClick={[Function]}
          value="TESTB"
        >
          TESTB
          <i
            aria-hidden="true"
            className="delete icon"
            onClick={[Function]}
          />
        </a>
        <a
          className="ui label"
          onClick={[Function]}
          value="TESTC"
        >
          TESTC
          <i
            aria-hidden="true"
            className="delete icon"
            onClick={[Function]}
          />
        </a>
        <input
          aria-autocomplete="list"
          autoComplete="off"
          className="search"
          onChange={[Function]}
          style={
            Object {
              "width": undefined,
            }
          }
          tabIndex={0}
          type="text"
          value=""
        />
        <span
          className="sizer"
        />
        <div
          aria-live="polite"
          className="text"
          role="alert"
        />
        <i
          aria-hidden="true"
          className="dropdown icon"
          onClick={[Function]}
        />
        <div
          aria-multiselectable={true}
          className="menu transition"
          role="listbox"
        >
          <div
            className="message"
          >
            No results found.
          </div>
        </div>
      </div>
      <hr />
      <button
        className="ui green button"
        disabled={undefined}
        onClick={[Function]}
        role="button"
        tabIndex={undefined}
        type="submit"
      >
        Save
      </button>
    </form>
  </div>
</div>
`;
